relay-server/src/services/project.rs

```diff
@@ -932,9 +933,39 @@ impl Project {
             Ok(current_limits.check_with_quotas(quotas, item_scoping))
         });
 
-        let (enforcement, mut rate_limits) =
+        let (mut enforcement, mut rate_limits) =
             envelope_limiter.enforce(envelope.envelope_mut(), &scoping)?;
-        enforcement.track_outcomes(envelope.envelope(), &scoping, outcome_aggregator);
+
+        #[derive(Debug, Deserialize)]
+        struct PartialEvent {
+            spans: SeqCount,
+        }
+
+        let check_nested_spans = state
+            .as_ref()
+            .is_some_and(|s| s.has_feature(Feature::ExtractSpansFromEvent));
+
+        // We compute the number of nested spans before rate limiting since we will remove the
+        // transaction item in case it is rate limited.
+        let spans = if check_nested_spans {
+            envelope
+                .envelope()
+                .items()
+                .find(|i| *i.ty() == ItemType::Transaction && !i.spans_extracted())
+                .and_then(|i| serde_json::from_slice::<PartialEvent>(&i.payload()).ok())
+                // We do + 1, since we count the transaction itself because it will be extracted
+                // as a span and counted during the slow path of rate limiting.
+                .map_or(0, |p| p.spans.0 + 1)
+        } else {
+            0
+        };
+
+        if spans > 0 {
+            add_nested_span_enforcements(&mut enforcement, spans);
+        }
+
+        enforcement.apply_with_outcomes(envelope.envelope_mut(), &scoping, outcome_aggregator);
```

Comment:
nit: Could you move all of this into a helper function? Maybe that helper function can then inline add_nested_span_enforcements.